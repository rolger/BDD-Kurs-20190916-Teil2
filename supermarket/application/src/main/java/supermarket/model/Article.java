package supermarket.model;

public class Article {
    private final ProductCode id;
    private final double quantity;

    public Article(ProductCode id, double quantity) {
        this.id = id;
        this.quantity = quantity;
    }

    public ProductCode getId() {
        return id;
    }

    public double getQuantity() {
        return quantity;
    }

}
