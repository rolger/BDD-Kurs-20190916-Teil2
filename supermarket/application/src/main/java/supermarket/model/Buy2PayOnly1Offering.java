package supermarket.model;

public class Buy2PayOnly1Offering implements Offering {

    @Override
    public Money calculateDiscount(Money basePrice, Article article) {
        int pieces = (int)article.getQuantity();
        int mod = pieces /2;
        return basePrice.multiply(mod).multiply(-1);
    }

    @Override
    public String getText() {
        return "  buy 2 pay 1";
    }

    @Override
    public boolean matchesWithProductUnit(Unit productUnit) {
        return Unit.PIECE.equals(productUnit);
    }

}
