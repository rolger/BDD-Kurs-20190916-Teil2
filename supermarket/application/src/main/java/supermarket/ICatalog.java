package supermarket;

import supermarket.model.Product;
import supermarket.model.ProductCode;

public interface ICatalog {

    public Product getProductByNumber(final ProductCode number);

}
