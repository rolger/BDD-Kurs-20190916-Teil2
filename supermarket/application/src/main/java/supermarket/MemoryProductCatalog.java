package supermarket;

import supermarket.model.Product;
import supermarket.model.ProductCode;

import java.util.Collection;
import java.util.HashMap;


public class MemoryProductCatalog implements ICatalog {
    private HashMap<ProductCode, Product> products = new HashMap<>();

    public Product getProductByNumber(final ProductCode number) {
        if (number == null) {
            throw new IllegalArgumentException("The ProductCode must not be null.");
        }
        return products.get(number);
    }

    public void add(Product product) {
        products.put(product.getCode(), product);
    }

    public void addAll(Collection<Product> additionalProducts) {
        additionalProducts.forEach(this::add);
    }

}