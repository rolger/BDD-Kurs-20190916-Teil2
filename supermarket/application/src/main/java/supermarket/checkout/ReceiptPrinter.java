package supermarket.checkout;

import supermarket.model.Receipt;
import supermarket.model.ReceiptItem;
import supermarket.model.Unit;

import java.time.LocalDate;
import java.util.Locale;

import static java.lang.String.format;
import static org.apache.commons.lang3.StringUtils.repeat;

public class ReceiptPrinter {

    public static final int COLUMNS = 40;
    public static final String TOTAL_TITLE = "Total :";

    private final LocalDate receiptDate;
    private final String shopName;

    public ReceiptPrinter(String shopName, LocalDate receiptDate) {
        this.shopName = shopName;
        this.receiptDate = receiptDate;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder receiptBuilder = new StringBuilder();

        printHeader(receiptBuilder);
        printLines(receipt, receiptBuilder);
        printFooter(receipt, receiptBuilder);

        return receiptBuilder.toString();
    }

    private void printHeader(StringBuilder receiptBuilder) {
        receiptBuilder.append(withCenterPadding(shopName));
        nextLine(receiptBuilder);
        nextLine(receiptBuilder);
        receiptBuilder.append("Your purchase from " + receiptDate);
        nextLine(receiptBuilder);
        separatorLine(receiptBuilder);
    }

    private void printLines(Receipt receipt, StringBuilder receiptBuilder) {
        receiptBuilder.append(withLeftPadding(0, "EUR"));
        nextLine(receiptBuilder);

        for (ReceiptItem item : receipt.getItems()) {
            String itemLine = formatLine(item);
            receiptBuilder.append(itemLine);
            receiptBuilder.append(withLeftPadding(itemLine.length(), item.getTotalPrice().format()));
            nextLine(receiptBuilder);
        }
    }

    private String formatLine(ReceiptItem item) {
        String itemLine = item.getText();
        if (shouldQuantityBePrinted(item)) {
            itemLine += " x ";
            itemLine += formatQuantity(item);
            itemLine += " ";
            itemLine += formatUnits(item);
        }
        return itemLine;
    }

    private boolean shouldQuantityBePrinted(ReceiptItem item) {
        return item.getQuantity() > 0 && item.getUnit() != null;
    }

    private String formatQuantity(ReceiptItem item) {
        return item.getUnit() == Unit.KG ?
                format(Locale.GERMAN, "%s", item.getQuantity()) :
                format(Locale.GERMAN, "%d", (int) item.getQuantity());
    }

    private String formatUnits(ReceiptItem item) {
        return item.getUnit() == Unit.KG ? "kg" : "";
    }

    private void printFooter(Receipt receipt, StringBuilder receiptBuilder) {
        separatorLine(receiptBuilder);
        receiptBuilder.append(TOTAL_TITLE);
        receiptBuilder.append(withLeftPadding(TOTAL_TITLE.length(), receipt.getTotalPrice().format()));
        nextLine(receiptBuilder);
    }

    private void separatorLine(StringBuilder receiptBuilder) {
        receiptBuilder.append(repeat("-", COLUMNS));
        nextLine(receiptBuilder);
    }

    private void nextLine(StringBuilder receiptBuilder) {
        receiptBuilder.append("\n");
    }

    private String withCenterPadding(String string) {
        return repeat(" ", (COLUMNS - string.length()) / 2) + string;
    }

    private String withLeftPadding(int length, String string) {
        return format("%1$" + (COLUMNS - length) + "s", string);
    }

}
