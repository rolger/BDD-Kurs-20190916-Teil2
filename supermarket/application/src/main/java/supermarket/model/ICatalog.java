package supermarket.model;

import supermarket.model.Product;
import supermarket.model.ProductCode;

public interface ICatalog {

    Product getProductByNumber(final ProductCode number);

}
