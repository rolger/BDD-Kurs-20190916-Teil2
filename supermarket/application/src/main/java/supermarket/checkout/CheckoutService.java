package supermarket.checkout;

import supermarket.model.ICatalog;
import supermarket.model.Product;
import supermarket.model.ProductCode;
import supermarket.model.Receipt;
import supermarket.model.ShoppingCart;

import java.time.LocalDate;

import static java.lang.String.format;

public class CheckoutService {
    private final String shopName;
    private final ICatalog catalog;
    private final ShoppingCart shoppingCart;

    public CheckoutService(String shopName, ICatalog catalog, ShoppingCart theCart) {
        this.shopName = shopName;
        this.catalog = catalog;
        this.shoppingCart = theCart;
    }

    public void onBarCode(String barcode, double quantity) {
        ProductCode code = new ProductCode(barcode);

        checkIfProductExists(barcode, code);
        shoppingCart.addArticle(code, quantity);
    }

    private void checkIfProductExists(String barcode, ProductCode code) {
        final Product product = catalog.getProductByNumber(code);
        if (product == null) {
            throw new IllegalArgumentException(format("The product with code %s doesn't exist in the catalog.", barcode));
        }
    }

    public String printReceipt(final CashRegister cashRegister, LocalDate receiptDate) {
        Receipt receipt = cashRegister.checkoutArticlesFrom(shoppingCart);
        return new ReceiptPrinter(shopName, receiptDate).printReceipt(receipt);
    }

}
