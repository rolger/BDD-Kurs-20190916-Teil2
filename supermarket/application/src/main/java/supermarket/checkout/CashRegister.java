package supermarket.checkout;

import supermarket.MemoryProductCatalog;
import supermarket.model.*;

import java.util.List;

public class CashRegister {

    private final MemoryProductCatalog catalog;

    public CashRegister(MemoryProductCatalog catalog) {
        this.catalog = catalog;
    }

    public Receipt checksOutArticlesFrom(ShoppingCart theCart) {
        Receipt receipt = new Receipt();

        List<Article> articles = theCart.getArticles();
        for (Article article : articles) {
            final Product product = catalog.getProductByNumber(article.getId());
            Money price = product.getBasePrice().multiply(article.getQuantity());
            receipt.addItem(product, article.getQuantity(), price);
        }

        return receipt;
    }

}
