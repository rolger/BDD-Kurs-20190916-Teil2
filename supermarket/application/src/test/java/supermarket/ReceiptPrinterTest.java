package supermarket;

import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Test;
import supermarket.checkout.CashRegister;
import supermarket.model.*;

import java.time.LocalDate;

import static org.approvaltests.Approvals.verify;
import static supermarket.model.ProductCode.buildNew;

public class ReceiptPrinterTest {

    private static final ProductCode apples = buildNew("00001");
    private static final ProductCode bread = buildNew("00002");
    private static final ProductCode milk = buildNew("00003");

    private static final LocalDate TRAINING_DATE = LocalDate.of(2019, 9, 17);

    private MemoryProductCatalog catalog;

    @Before
    public void setUp() throws Exception {
        catalog = new MemoryProductCatalog();
        catalog.add(new Product(apples, "Apfel", new Money(0.9)));
        catalog.add(new Product(bread, "Schwarzbrot", new Money(2.49)));
        catalog.add(new Product(milk, "Milch", new Money(1.01)));
    }

    @Test
    public void buyOneArticle() {
        ShoppingCart aCart = new ShoppingCart();
        aCart.addArticle(bread, 2);

        Receipt r = new CashRegister(catalog).checksOutArticlesFrom(aCart);

        verify(new ReceiptPrinter(TRAINING_DATE).printReceipt(r));
    }
    @Test
    public void buyThreeArticlesWithoutDiscount() {
        ShoppingCart aCart = new ShoppingCart();
        aCart.addArticle(apples, 0.75);
        aCart.addArticle(bread, 1);
        aCart.addArticle(milk, 1);

        Receipt r = new CashRegister(catalog).checksOutArticlesFrom(aCart);

        verify(new ReceiptPrinter(TRAINING_DATE).printReceipt(r));
    }
}
