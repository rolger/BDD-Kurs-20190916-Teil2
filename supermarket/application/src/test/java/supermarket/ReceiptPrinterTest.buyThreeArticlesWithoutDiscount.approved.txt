This is your receipt from 2019-09-17

                                     EUR
Apfel                               0,68
Schwarzbrot                         2,49
Milch                               1,01

Total:                              4,18
